### ISSUE TITLE
Publish blog post about integrating Kedro and Ibis

### ISSUE THREAD

**Issue by deepyaman** [kedro-org/kedro-devrel#121]
# Objectives

Wanted to share something I've been wanting to do for a while, and finally got around to doing in the past couple weeks: show how Kedro and Ibis can be used together as a Python-first alternative to dbt.

Here's a Kedro project implementing the prototypical Jaffle Shop example: https://github.com/deepyaman/jaffle-shop If you want to try it out, you should be able to install dependencies from `requirements.txt` in a virtual environment (it's currently using Kedro off the development branch, but I'll change that to use 0.19), run `kedro run --pipeline seed` to set up the DuckDB database, and `kedro run` from there.

I think there are a lot of Kedro projects where something like this could help a lot/be the "right" thing to do. Just from my experience using Kedro at clients, I know there are many cases where people:

* slurp up large amounts of data into memory, instead of pushing execution down to the database/engine
* implement the 3-4 month PoC on data extracts, and then struggle massively when needing to migrate to running against the production databases + scale out
* insist on using Kedro across the full workflow for consistency (fair) when dbt would have been the much better fit, because you essentially needed a SQL workflow

I think Kedro and Ibis fit well together, too, because Kedro is "production-ready data pipelines using Python" and Ibis also makes the dev-to-prod transition easier using a Python dataframe API.

# Post

## Candidate titles

* _From production-ready to production: building scalable data pipelines with Kedro and Ibis_
  * Aligned with the purpose of each library
    * "Kedro is a toolbox for production-ready data science."
    * Ibis provides "the flexibility of Python with the scale and performance of modern SQL."
  * Alternative versions
    * _From production-ready to production: building scalable Kedro pipelines with Ibis_
      * I think I like this less, because building scalable data pipelines is, by definition, interesting to a broader audience than building scalable Kedro pipelines
* _Back to the Jaffle Shop! A Python-first implementation of the prototypical dbt project_
  * "dbt™ is a SQL-first transformation workflow that lets teams quickly and collaboratively deploy analytics code following software engineering best practices," which is very similar to what Kedro does for Python users
    * We're not competing (IMO), but maybe dropping dbt in the title is too close for @astrojuanlu
  * Alternative versions
    * _Welcome to the Tosti Shop!_
      * Because tostis are Dutch waffles 😝 I'm pretty sure 50% of the readership won't get that, and the other 50% won't read it because of the bad joke

## Outline

* **The dev-prod dilemma**
  * (As a Kedro user) Kedro excels in the PoC/development phase
    * Kedro helps data engineers, data scientists, and machine learning engineers collaboratively build end-to-end data pipelines, without throwing software engineering out the window
    * (From personal experience) During development, usually work with data extracts--CSVs, dev databases, even Excel files 🤮 
    * Kedro's connector ecosystem (Kedro-Datasets) makes it easy
  * However, deploying the same code in production doesn't work as well as one would hope...
    * (From personal experience) Often times, the solution doesn't scale
      * A lot of teams preemptively leverage (Py)Spark for their data engineering workfloads, but Spark often underperforms code directly executed on the backend engine (and that's assuming you take advantage of predicate pushdown)
    * Machine learning (software) engineer rewrites the code, but...
      * It takes (a lot of) time, and the business doesn't understand why--it worked before!
      * Rewritten logic, intentionally or not, doesn't always correspond 1:1 to the original, leading to unexpected results (read: 🐛)
* **The SQL solution**
  * Thankfully, there's a standardized* programming language that's supported by every database (and many other compute frameworks)
    * Can put a footnote linking out to Phillip or Gil's talk, on how SQL isn't really standardized; even the [dbt-labs/jaffle_shop](https://github.com/dbt-labs/jaffle_shop) repo says, "If this steps fails, it might mean that you need to make small changes to the SQL in the models folder to adjust for the flavor of SQL of your target database."
  * _If you like SQL,_ dbt provides a battle-tested 
    * The Kedro team loves dbt! In fact, we've frequently recommended it ELT workflows
    * We're all about empowering data teams to write ["reproducible, maintainable, and modular"](https://docs.kedro.org/en/stable/introduction/index.html#introduction-to-kedro) code
    * Shared goals: "dbt™ is a SQL-first transformation workflow that lets teams quickly and collaboratively deploy analytics code following software engineering best practices," which is very similar to what Kedro does for Python users
* **"What if I don't like SQL?"**
  * Many teams (and data scientists) are already using Kedro and Python for their data science workflows
  * "Ibis is a Python library that provides a lightweight, universal interface for data wrangling"
    * A dataframe API for Python
    * Interfaces to 15+ query engines
    * Deferred execution
* **Revisiting the Jaffle Shop**
  * If you're not familiar with the Jaffle Shop, it is the... [brief description of the Jaffle Shop project]
  * Since dbt is the industry standard for SQL-first transformation workflows, let's use the Jaffle Shop example to inform what we should be able to do in a Python-first solution
  * (Provide option to skip ahead to final solution using the started, for people who just want to see it in action)
  * Creating a custom `ibis.Table` dataset
  * Configuring backends using the `OmegaConfigLoader` using variable interpolation
  * Building the pipelines
    * Walk through `process_orders` node implementation (because it also showcases parametrization)
  * Putting it all together
  * Quickstart: `kedro-starter-jaffle-shop`
* **But wait... there's more!** (tangential benefits)
  * Make backend (database) migration easy
    * (From personal experience) sometimes you're developing against the existing data system, even though you know the data infrastructure team is currently working on a migration to a new set of databases, and your code won't work exactly... unless you use Ibis?
  * Build truly reusable pipelines
    * (From personal experience) leading the development of an asset that provided a suite of reusable pipelines for customer analytics; we made the decision to use Spark for data engineering, but...
      * One major retailer had all their data in Oracle, and they ended up with a suboptimal workflow wherein they extracted the data into Spark and did all the work there
      * A lot of teams looked at our pipelines for inspiration, but ended up rewriting them for their environment--not the intention
      * We spent so much time and effort setting up Spark on locked-down Windows machines, so data scientists could play with the pipelines; it would have been so much easier to run the same logic locally using DuckDB or pandas instead
* **What's next?**
  * Validations

# Action items

- [x] Update https://github.com/deepyaman/jaffle-shop to use (now-released) Kedro 0.19 instead of `develop` build
- [ ] Create a Kedro starter, so people can try the project themselves easily
- [ ] Add `ibis.TableDataset` to `kedro-datasets` (not a prerequisite for the post; I'd like to walk through the dataset implementation at a high level anyway)
- [ ] Add data validation capabilities using Kedro-Pandera (for a part 2 post; will require additional work like extending Pandera)

**Comment by stichbury** [comment]
This is awesome and I'd definitely like to work on a post with you. Please let me know when/if/how you want to proceed.

**Comment by deepyaman** [comment]
> This is awesome and I'd definitely like to work on a post with you. Please let me know when/if/how you want to proceed.

I'll draft the post based on the above proposed outline (any feedback welcome), and then will of course ask you to review! Aiming to have this ready to publish within the next month, so will write the content first draft in the next week or two.

**Comment by merelcht** [comment]
> Welcome to the Tosti Shop!
Because tostis are Dutch waffles 😝 I'm pretty sure 50% of the readership won't get that, and the other 50% won't read it because of the bad joke

Are they? A tosti in dutch is a toasted cheese sandwich, like a croque monsieur. A dutch waffle is a stroopwafel IMO 😄 

Jokes aside, I think this would be a really fantastic post for the Kedro blog! I'd be more than happy to review as well!

**Comment by deepyaman** [comment]
> > Welcome to the Tosti Shop!
> > Because tostis are Dutch waffles 😝 I'm pretty sure 50% of the readership won't get that, and the other 50% won't read it because of the bad joke
> 
> Are they? A tosti in dutch is a toasted cheese sandwich, like a croque monsieur. When you say "a Dutch waffle" I think of a stroopwafel 😄

@merelcht Autocorrect got me! I meant, "tostis are Dutch jaffles." You can fact-check that statement. :)

**Comment by merelcht** [comment]
> @merelcht Autocorrect got me! I meant, "tostis are Dutch jaffles." You can fact-check that statement. :)

Haha! Yes that is indeed true 😄 🍞🧀

**Comment by stichbury** [comment]
We've had some comments here https://linen-slack.kedro.org/t/16361361/hi-u03s12lhnnq-u03r8fw4huz-u046v0qc124-fantastic-https-kedro#4c6fdd09-13dc-4f21-99de-25dec267a3a4 and it sounds as if a follow-up post could be valuable, particularly as this is proving a popular topic.

@deepyaman would you be able to summarise the thread in a new issue, or shall I do it and let you correct it when I've messed up the argument?

**Comment by conradbez** [comment]
Might not be the right place but I've had similar issues but found dbt was overall a better solution for most data wrangling tasks now that there is python model support.  Kedro seems more suitable when it comes to testing many different ML model skews though.

I created this library to ease my transition from Kedro to dbt and provide more of Kedro data swiss army knife abilities to dbt: https://github.com/conradbez/dbt-duckdb-kedro-datasets

Basically it allows pasting Kedro data catalog directly into dbt source files and ingesting tabular data just like you would in Kedro. I find it super easy to then delete the `pipelines.py` file in in Kedro and create dbt python models which import and use the logic from the existing `nodes.py` files.

**Comment by astrojuanlu** [comment]
Hi @conradbez, since you're here already, thanks for sharing! What's your motivation for using Kedro + dbt? Would you like to share in a new Discussion https://github.com/kedro-org/kedro/discussions/ ?
